### ISSUE TITLE
[clang] False Positive in Wunneded-member-function?

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#52835]
The following code triggers a potential false positive in `Wunneeded-member-function`

```
namespace 
{
struct Foo
{
    Foo() = default;
    
    Foo(Foo const&) = delete;
    Foo& operator=(Foo const&) = delete;

    Foo(Foo&&) = default;
    Foo& operator=(Foo&&) = default;
};
}

void f(Foo f) {}


int main()
{
    f(Foo());
}
```
```
source>:5:5: warning: member function 'Foo' is not needed and will not be emitted [-Wunneeded-member-function]
    Foo() = default;
    ^
```

If I remove the constructor, the code won't compile.

Repro: https://godbolt.org/z/cha1sqE7q

**Comment by shafik** [comment]
dup: https://github.com/llvm/llvm-project/issues/42943

**Comment by aaronpuchert** [comment]
I don't think this is a duplicate of #42943. As @zygoloid pointed out there, a definition is not needed in that case, only a declaration. However, here we need the full definition.

I think it might be the same issue as #29187 though: a member function where we need a full definition (because the function is actually being called), but where the member function is a trivial default constructor or destructor, so a call does not need to be emitted.
